desktop/renderer-app/scripts/vite-plugin-electron.ts

```diff
@@ -0,0 +1,174 @@
+
+const needParse = (
+    id: string,
+    entries: Array<{ entryPoints: string; code: string; shouldBundle: boolean }>,
+): Promise<string | null> => {
+    const cid = cleanUrl(id);
+    let result: Promise<string | null> = Promise.resolve(null);
+    for (const key of entries) {
+        if (cid.endsWith(key.entryPoints)) {
+            if (key.shouldBundle) {
+                const path = id.replace(/(\?v=)(\w+)/g, "");
+                result = build({
+                    entryPoints: [path],
+                    bundle: true,
+                    outfile: "index.js",
+                    platform: "node",
+                    write: false,
+                })
+                    .then(result => {
+                        return (
+                            result?.outputFiles
+                                ?.map(item => {
+                                    return item.text;
+                                })
+                                .join("") ?? ""
+                        );
+                    })
+                    .then(result => {
+                        return result.concat(key?.code);
+                    })
+                    .catch(e => {
+                        console.warn(e);
+                        return null;
+                    });
+            } else {
+                return Promise.resolve(key.code);
+            }
+        }
+    }
+    return result;
+};
+/**
+ * The externals will be work on 'viteConfig.build.rollupOptions.external', 'viteConfig.optimizeDeps.exclude' option.
+ * and we should external ['electron', ...require('module').builtinModules]
+ */
+const externals = ["electron", ...builtinModules];
+export function electron(): VitePlugin[] {
+    const nodeModelResolve: VitePlugin = {
+        name: "vite-plugin-electron:node-model-resolve",
+        apply: "serve",
+        transform(_code, id) {
+            return needParse(id, entries);
+        },
+    };
+
+    const builtinModulesResolve: VitePlugin = {
+        apply: "serve",
+        name: "vite-plugin-electron:builtinModules-resolve",
+        transform(_code, id) {
+            if (id.includes("__vite-browser-external")) {
+                const moduleId = id.split(":")[1];
+                if (builtinModules.includes(moduleId)) {
+                    const nodeModule = require(moduleId);
+                    const attrs = Object.keys(nodeModule);
+                    const requireTpl = `const __nodeModule = require('${moduleId}');`;
+                    const declaresTpl =
+                        attrs.map(attr => `const ${attr} = __nodeModule.${attr}`).join(";\n") + ";";
+                    const exportTpl = `export {\n  ${attrs.join(",\n  ")},\n}`;
+                    const exportDefault = `export default { ${attrs.join(", ")} };`;
+
+                    const nodeModuleCode = `
+${requireTpl}
+${declaresTpl}
+${exportTpl}
+${exportDefault}
+`;
```

Comment:
Use `['', requireTpl, declaresTpl, exportTpl, exportDefault, ''].join('\n')` will be better.